Posts: Add single site post requesting behavior to Redux state

aduth opened the pull request:
This pull request seeks to implement Redux state behaviors for managing network requests of single site post objects. This will be used as part of the editor Redux migration effort in allowing the editor to reconcile post revisions with the persisted post entity.

__Testing instructions:__

There are no effective visual changes to test.

Ensure Mocha tests pass by running `make test` in your Terminal from the project root directory or directly from `client/state/posts`.

blowery commented:
Test pass locally, woo

blowery commented on client/state/action-types.js:
```diff
@@ -23,6 +23,9 @@ export const FETCH_SITE_PLANS = 'FETCH_SITE_PLANS';
 export const NEW_NOTICE = 'NEW_NOTICE';
+export const POST_REQUEST = 'POST_REQUEST';
+export const POST_REQUEST_SUCCESS = 'POST_REQUEST_SUCCESS';
+export const POST_REQUEST_FAILURE = 'POST_REQUEST_FAILURE';
```
(note: not really for this PR)

I wonder if we should flip all of these to keymirror... It turns out the const doesn't transfer through the export, so an importer can still reassign the imported variable. I feel like the `const` is giving a false sense of security. 

But maybe it helps with code completion in some editors?

aduth replied:
>I wonder if we should flip all of these to keymirror

What advantage would there be in using keyMirror?

For me at least, this is less about feeling of security with `const` and rather in opting to use simple language constructs than an external library that, in my opinion, doesn't provide much value.

There's some discussion in the `keyMirror` repository [about minifiers crushing keys](https://github.com/STRML/keyMirror#why), though I don't think it affects our situation.

blowery replied:
just less duplication. it's really easy to fat finger one side of the const construction and not notice.

MY_REALLY_LONG_KEY = "MY_REALY_LONG_KEY"

vs

keyMirror( { MY_REALLY_LONG_KEY: t } )

aduth replied:
>it's really easy to fat finger one side of the const construction and not notice

Maybe some custom lint rules could help here?

Related: #3015, https://github.com/Automattic/eslint-plugin-wpcalypso

blowery commented on client/state/posts/actions.js:
```diff
@@ -68,3 +71,36 @@ export function requestSitePosts( siteId, query = {} ) {
+				siteId,
+				postId
+			} );
+		} ).catch( ( error ) => {
```
(this is curiosity talking, not a criticism) Why use `catch` instead of passing the error handler as the second arg to `then`?

aduth replied:
>Why use `catch` instead of passing the error handler as the second arg to `then`?

Don't feel strongly either way, but I suppose an argument could be made that it lends more clarity to what the code is effectively accomplishing, as `catch` is more meaningful than happening to be passed as the second argument.

blowery replied:
There's a real subtle diff between the two cases. By using catch after the then, you're also catching any errors that the first then handler throws. If you use the `then( yay, boo )` construct, `boo` will never see errors that `yay` throws. So it really depends on what you're trying to capture and who's going to consume the Promise.

see http://www.html5rocks.com/en/tutorials/es6/promises/#toc-error-handling

aduth replied:
>There's a real subtle diff between the two cases.

Subtle indeed, though still unsure which makes the most sense. Would we want to allow the `yay` case to throw uncaptured, or is it preferable to have the catch block capture errors (dispatching failure action) in both the request and `yay` handler? Obviously we don't want errors in our handlers in any situation, so perhaps not important to dwell too much over.

blowery replied:
Yup, don't dwell. Usage here is fine. If you want to flip to returning the promise from the test, just drop the catch.

blowery commented on client/state/posts/actions.js:
```diff
@@ -68,3 +71,36 @@ export function requestSitePosts( siteId, query = {} ) {
+				type: POST_REQUEST_SUCCESS,
+				siteId,
+				postId
+			} );
```
this should probably return the post in case someone consuming this wants to chain onto the Promise and use it. See http://jsbin.com/jufayefego/edit?js,console

aduth replied:
>this should probably return the post in case someone consuming this wants to chain onto the Promise and use it. See http://jsbin.com/jufayefego/edit?js,console

Arguably, we shouldn't want to allow this, as the only consumption of data should be directly from the Redux state tree, not from an action thunk resolution. The only reason I opted to return the Promise in the first place is in facilitating test's awareness of the requests completing (as nock mocks aren't as immediate as you might expect). It may also come in handy for server-side rendering in knowing whether the response is ready to serve (based on the resolution of all initial actions), but again, not in specifically consuming any data result.

blowery replied:
> as nock mocks aren't as immediate as you might expect
Promises always fulfill or reject on the next (well, a future) turn of the event loop, unless you mean something else? See http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony for most of the reasoning. Zalgo is bad.

> Arguably, we shouldn't want to allow this, as the only consumption of data should be directly from the Redux state tree, not from an action thunk resolution.

I can get behind that.

aduth replied:
>Promises always fulfill or reject on the next (well, a future) turn of the event loop, unless you mean something else?

Yeah, I mean something else. Naive attempts by me in testing nock would look more like:

```jsx
it( 'should finish request on next tick', ( done ) => {
	someRequestMockedByNock();

	setTimeout( () => {
		expect( myAsyncSpy ).to.have.been.calledOnce;
		done();
	}, 0 );
} );
```

But for whatever reason, the nock mocks aren't resolved in any predictable next tick.

blowery replied:
Ah, interesting! Good to know.

blowery commented on client/state/posts/selectors.js:
```diff
@@ -158,3 +158,20 @@ export function getSitePostsForQueryIgnoringPage( state, siteId, query ) {
+ * @return {Boolean}        Whether request is in progress
+ */
+export function isRequestingSitePost( state, siteId, postId ) {
+	if ( ! state.posts.siteRequests[ siteId ] ) {
```
are you guaranteed that `state.posts.siteRequests` exists?

aduth replied:
>are you guaranteed that `state.posts.siteRequests` exists?

Yep, because the Redux store will be initialized with the [default return value of the reducer](https://github.com/Automattic/wp-calypso/blob/615773ac9f6c4c9d596ae3a0e990d80278d3572e/client/state/posts/reducer.js#L54). Under the hood, the initial state is constructed during the Redux internal `@@INIT` action type.

blowery replied:
> Yep, because the Redux store will be initialized with the default return value of the reducer. 

Are you guaranteed a caller won't do run this before that happens?

aduth replied:
>Are you guaranteed a caller won't do run `isRequestingSitePost` before redux initializes?

Yes, or at least, if there are callers, that's a larger issue in itself.

blowery replied:
Good good

blowery commented on client/state/posts/test/actions.js:
```diff
@@ -153,4 +163,70 @@ describe( 'actions', () => {
+
+	describe( '#requestSitePost()', () => {
+		before( () => {
+			nock( 'https://public-api.wordpress.com:443' )
```
not for this PR, but if it's too painful to mock wpcom.js, maybe we should look at fixing wpcom.js to make it easier to test through. Having to bake all this info (host / port / protocol / path) into tests feels fragile.

aduth replied:
>maybe we should look at fixing wpcom.js to make it easier to test through.

Yeah, I've used and seen a few different approaches to mocking WPCOM.js requests, though I'm mostly content with this one (see also: [mocking WPCOM.js methods gets hairy too](https://github.com/Automattic/wp-calypso/blob/b29d2b1c31642e8f34a1b7f2491401af5d35b62b/client/lib/media/test/actions.js#L56-L71)). 

Maybe some abstractions or constants to reduce the chance of changes breaking tests? I dunno, seems a bit premature, and this doesn't feel particularly troubling to me.

blowery commented on client/state/posts/test/actions.js:
```diff
@@ -153,4 +163,70 @@ describe( 'actions', () => {
+		} );
+
+		it( 'should dispatch posts receive action when request completes', ( done ) => {
+			requestSitePost( 2916284, 413 )( spy ).then( () => {
```
i would expect this then success callback to have been called with the post too

blowery replied:
n/m as discussed before

blowery commented on client/state/posts/test/actions.js:
```diff
@@ -153,4 +163,70 @@ describe( 'actions', () => {
+			} );
+		} );
+
+		it( 'should dispatch posts receive action when request completes', ( done ) => {
```
general thing, but i'm pretty sure you can just return the promise from requestSitePost.then(...) instead of using the done callback. https://mochajs.org/#working-with-promises

aduth replied:
>i'm pretty sure you can just return the promise

Ooh, TIL! That's very handy indeed. Changed in fbb511f.

blowery commented on client/state/posts/test/actions.js:
```diff
@@ -153,4 +163,70 @@ describe( 'actions', () => {
+				} );
+
+				done();
+			} ).catch( done );
```
i would expect the catch to be called with the error action

aduth replied:
>i would expect the catch to be called with the error action
 
That would assume that the failing case would throw an error in the client, which is undesirable (should be handled by Redux state if we want to handle at all).

The `catch` in the tests was more to catch any runtime errors during the test that would otherwise cause the tests to obscurely fail with a timeout. Mostly moot now that I'm returning the promise in fbb511f.

blowery replied:
in light of the previous discussion, this is ... odd now. Do we want the Promise to resolve or reject if the API request bombs?

aduth replied:
>Do we want the Promise to resolve or reject if the API request bombs?

We want the promise to resolve successfully regardless of whether an API failure occurs. API failures are handled by the `POST_REQUEST_FAILURE` action.

blowery commented on client/state/posts/test/actions.js:
```diff
@@ -153,4 +163,70 @@ describe( 'actions', () => {
+
+		it( 'should dispatch fail action when request fails', ( done ) => {
+			requestSitePost( 2916284, 420 )( spy ).then( () => {
+				expect( spy ).to.have.been.calledWith( {
```
i wouldn't expect the success callback to be called at all?

aduth replied:
>i wouldn't expect the success callback to be called at all?

Similar points as above in https://github.com/Automattic/wp-calypso/pull/3108#discussion_r52024896 and https://github.com/Automattic/wp-calypso/pull/3108#discussion_r52023037. The promise really only exists to mark progress of the request, but is not intended to be used in any effective way (either through consumption of the resolved object, or in throwing errors that occurred during the request). These are better handled by the Redux action handlers.

blowery commented on client/state/posts/test/reducer.js:
```diff
@@ -366,4 +370,105 @@ describe( 'reducer', () => {
+		} );
+
+		it( 'should accumulate mappings', () => {
+			const state = siteRequests( Object.freeze( {
```
I think you need to deepfreeze to really verify, as the thing being mutated is a level down

aduth replied:
>I think you need to deepfreeze to really verify, as the thing being mutated is a level down

Ooh, wow, you're totally right, and there are existing reducers which are breaking because of the failure of `Object.assign` to deeply clone objects too. The [`siteQueries` reducer](https://github.com/Automattic/wp-calypso/blob/b29d2b1c31642e8f34a1b7f2491401af5d35b62b/client/state/posts/reducer.js#L80-L104), for example, suffers from an issue similar to http://jsbin.com/puyuhepiju/edit?html,js,console .

And before any mention of Immutable.js is made, the use of `deepFreeze` in tests is effective in ensuring that these are properly handled :wink:

blowery commented on client/state/posts/test/reducer.js:
```diff
@@ -366,4 +370,105 @@ describe( 'reducer', () => {
+		} );
+
+		it( 'should map site ID, post ID to false value if request finishes successfully', () => {
+			const state = siteRequests( Object.freeze( {
```
same on the need to deepfreeze

blowery commented on client/state/posts/test/reducer.js:
```diff
@@ -366,4 +370,105 @@ describe( 'reducer', () => {
+			} );
+		} );
+
+		it( 'should map site ID, post ID to false value if request finishes with failure', () => {
```
might want to keep around the reason it failed, at least for debugging? false doesn't tell us much.

aduth replied:
>might want to keep around the reason it failed, at least for debugging?

Depends on how the error would be used. I could see it being shown in a notice, for example, but that itself might be a separate reducer. This reducer is exclusively responsible for tracking the request progress.

blowery commented on client/state/posts/test/selectors.js:
```diff
@@ -361,4 +362,44 @@ describe( 'selectors', () => {
+			}, 2916284, 841 );
+
+			expect( isRequesting ).to.be.false;
+		} );
```
might want a test where posts.siteRequests doesn't exist

aduth replied:
>might want a test where posts.siteRequests doesn't exist

`siteRequests` is never unset, per https://github.com/Automattic/wp-calypso/pull/3108#discussion_r52023383

aduth commented:
Thanks for the tip on `deep-freeze`, @blowery . Introduced it in 52ad4f8, which unveiled a few bugs in the existing posts reducers, fixed in the same commit.

Will definitely plan on using `deep-freeze` consistently in future tests.

blowery commented:
Looks good, unleash the 🐲.

aduth commented:
Thanks for the review, @blowery ! Some great discussion/discoveries in this pull request :smile:
